modules/core/src/test/scala/org/tessellation/trust/SybilAttackSuite.scala

```diff
@@ -1,37 +1,134 @@
+
+  val random: util.Random = new scala.util.Random(0L)
+
+  def benchmark(allData: List[TrustNode], filterGood: Int => Boolean): ModelBenchmark = {
+
+    val numNodes = allData.size / 2
+    val scores = SelfAvoidingWalk.runWalkFeedbackUpdateSingleNode(
+      0,
+      allData
+    )
+    val relativeLocalTrust = scores.edges.filter(e => filterGood(e.dst)).map(_.trust).sum
+    val relativeMaliciousTrust = scores.edges.filter(e => !filterGood(e.dst)).map(_.trust).sum
+
+    val eigenTrustComparisonScores = EigenTrust.calculate(allData)
+    val eigenTrustDefenseFactor = eigenTrustComparisonScores.filter(e => filterGood(e._1)).values.sum /
+      eigenTrustComparisonScores.filter(e => !filterGood(e._1)).values.sum
+
+    val defensiveFactor = relativeLocalTrust / relativeMaliciousTrust
+
+    val dattScores = DATT.calculate(DATT.convert(allData), 0)
+
+    val dattDefenseFactor = dattScores.filter(e => filterGood(e._1)).values.sum /
+      dattScores.filter(e => !filterGood(e._1)).values.sum
+
+    val modelScore = TrustModel.calculateTrust(allData, 0)
+
+    val modelDefenseFactor = modelScore.filter(e => filterGood(e._1)).values.sum /
+      modelScore.filter(e => !filterGood(e._1)).values.sum
+
+    ModelBenchmark(
+      sawDefenseFactor = defensiveFactor,
+      dattDefenseFactor = dattDefenseFactor,
+      eigenTrustDefenseFactor = eigenTrustDefenseFactor,
+      modelDefenseFactor = modelDefenseFactor
+    )
+  }
+
+  def averageBenchmark(benchmarks: List[ModelBenchmark]): ModelBenchmark =
+    ModelBenchmark(
+      benchmarks.map(_.sawDefenseFactor).sum / benchmarks.size,
+      benchmarks.map(_.dattDefenseFactor).sum / benchmarks.size,
+      benchmarks.map(_.eigenTrustDefenseFactor).sum / benchmarks.size,
+      benchmarks.map(_.modelDefenseFactor).sum / benchmarks.size
+    )
+
+  test("Data generator") {
+
+    val numNodes = 100
+    val numNetworks = 10
 
     val generator = Random.scalaUtilRandomSeedInt[IO](0).map { implicit rnd =>
       new DataGenerator[IO]
     }
     for {
-      _ <- ignore("Non-deterministic unit test!").unlessA(false)
+      _ <- ignore("Comment to modify data generation").unlessA(false)
       gen <- generator
-      data <- gen.generateData(numNodes = numNodes, edgeLogic = gen.randomPositiveEdge()).map { nodes =>
-        nodes.map { node =>
+      networkData <- List
+        .fill(numNetworks)(gen.generateData(numNodes = numNodes, edgeLogic = gen.randomPositiveEdge()))
+        .foldLeft(IO.pure(List.empty[List[TrustNode]])) {
+          case (col, next) =>
+            next.flatMap { n =>
+              col.map(_ :+ n)
+            }
+        }
+    } yield {
+      val ser = networkData.flatMap { _.map { _.asJson.noSpaces } }
+      val res = decode[TrustNode](ser.head)
+      Files
+        .write(Paths.get("./src/test/resources/trust_100x10.json"), ser.mkString("\n").getBytes(StandardCharsets.UTF_8))
+      expect(res.isRight)
+    }
+  }
+
+  def loadSerialized(): IO[List[List[TrustNode]]] = {
```

Comment:
Your test was non-deterministic. It means that for given preconditions there are some values that could lead to a failure. I'd advise not to lock the test on the optimistic version (what you did right now) but to fix the algorithm to pass all the edge cases. Otherwise, you're introducing a test that checks if you created a correct json file, but doesn't check if the algorithm works properly for the given arbitrary data.

Reply:
Most of the non-determinism is coming from the old SAW module which I'm incrementally refactoring. The 'optimistic' version here should absolutely remain as a purely deterministic version -- it verifies given known data that we get precise outputs. DATT for instance is purely deterministic and has no randomness associated with it -- so fixed inputs are totally fine and yield exact outputs. The current SAW outputs -- even on fixed inputs, are not deterministic at all due to random seed issue which also needs to be fixed. 

That's one level of tests, but doesn't evaluate model performance as properly as possible. I actually want a second level of unit tests that covers more dynamically generated configurations -- but there's several edge cases here that require much more refactoring to eliminate (primarily due to the old SAW code which cannot be easily fixed immediately.) For now though, I don't want to introduce accidentally failing tests -- and these at least demonstrate some level of security which can be improved upon.

Evaluating the actual model and fixing hyper-parameters for the model is completely outside the scope of a unit test, and belongs as part of a model training framework. It will take too long (on the order of hours of compute time) to properly evaluate a model on many many network configurations. That is what we need to demonstrate proper security anyways -- any unit test is not going to capture enough configurations to demonstrate anything beyond a very basic proof that shows the code makes sense -- that's the only thing being shown here are a few attack scenarios to demonstrate reasons for current model.

Reply:
Main purpose of these tests also is to prevent any regressions while model is being developed -- future development should not substantially break these without manual oversight of changes in scores. Fully proving security is not complete, but this is a good place to start.